### ISSUE TITLE
[Proposal] Recipe for private repositories

### ISSUE THREAD

**Issue by francoispluchino** [symfony/flex#71]
Using a server (https://symfony.sh) to define the Recipes is an excellent initiative for open-source packages, allowing to create an automatic configuration for Symfony even if the author of the package didn't include / doesn't wish to include this configuration. It is a practice more and more used, in particular in the world of javascript, like [TypeScript type definitions](https://github.com/DefinitelyTyped/DefinitelyTyped) for example.

However, the repositories [symfony/recipes](https://github.com/symfony/recipes) and [symfony/recipes-contrib](https://github.com/symfony/recipes-contrib) are public, and require approval to add or modify their manifests and their configurations (manual and strict for `official`, semi-automatic if I understood correctly for `contrib`). This approach is therefore not suitable for private packages.

Questions are beginning to be asked about this on the Medium [Symfony 4: Contributing Recipes](https://medium.com/@fabpot/symfony-4-contributing-recipes-2f489dc6f645), With an approach to create a private server, And asking the question about opening the source code of symfony.sh to create this private server.

I'm not against this approach quite the opposite, it sounds like creating a Satis server for Composer. However, it would be simpler/quicker to add the manifest and configurations to a dedicated directory in the repository. Just like Composer, that we can add repositories that are not referenced in Packagist or a Satis private server by using the VCS Repository. But unlike the VCS Repositories of Composer, the performance are not impacted, quite the opposite (no HTTP request).

The path to the Recipe directory might will have a default value (`./recipe/` or other), with a possibility to change this value in the `composer.json` file with the `extra.symfony.recipe-path` key (or another name). If this directory exists in the repository with the `manifest.json` file, in this case, no request is done on the `symfony.sh` server.

This approach can be easily implemented given that the automatic configuration is performed just after the installation of the dependency and added just before the search of recipe on the symfony.sh server (in `Symfony\Flex\Flex::filterPackageNames()`.

**To summarize the workflow after the installation of package:**

1. Check the presence of the key `extra.symfony-recipe-path` in `composer.json` file of package
2. Get the real path of package recipe
3. Check if the `manifest.json` file is present and valid
4. If package recipe is valid, install the automatic configuration from package source code
5. Else, search recipe in symfony.sh server and install the automatic configuration from remote recipe if recipe is present on the server (the current behavior)

What do you think?

**Comment by dkarlovi** [comment]
Check #70 for their reasons why not.

**Comment by francoispluchino** [comment]
@dkarlovi I didn't see your issue because it was closed, and that I stopped at the open issues. Thanks!

**Comment by dkarlovi** [comment]
Currently, I see there's support to specify a custom endpoint in `extra.symfony.endpoint`, but as there's nothing to point to, you can't do anything about it. :)

**Comment by francoispluchino** [comment]
Yes but a flexible approach would be a plus, just like Composer offers:

1. Packagist
2. Satis private server
3. VCS Repository

So for Flex:

1. Recipes on https://symfony.sh
2. Recipes contrib on https://symfony.sh (require approval in composer.json file)
3. Recipes on private server (require approval in composer.json file)
4. Recipes in each package (require approval in composer.json file)

> Knowing that 1 and 2 are already implemented for Flex.

**Comment by dkarlovi** [comment]
@francoispluchino I very much agree, obviously. :+1: Maybe they change their mind on this if enough of people asks about it.

**Comment by fabpot** [comment]
Keep in mind that Flex is a preview of a pre-alpha :) It's something that will be stable for Symfony 4, at the end of November.

I'm very happy that many people are as excited as I am though! That makes me want to work even harder to make it stable ASAP.

Regarding storing recipes in the Github repo of the package itself, that won't work for many reasons, some explained here: https://github.com/symfony/recipes/issues/53#issuecomment-299081540.

**Comment by francoispluchino** [comment]
Another problem with `extra.symfony.endpoint`, if you change its value, you no longer have access to the recipes of https://symfony.sh, unless you create a proxy server for this private server.

**Comment by fabpot** [comment]
The `extra.symfony.endpoint` is only useful for testing PRs.

**Comment by francoispluchino** [comment]
@fabpot I know, it was a pseudo additional argument to support my proposition :)

Do not worry, I'm very well aware that this plugin is in Alpha, And just like you, I'm very excited about this feature, especially - But you probably don't remember it, and it's normal - Especially since I contacted you at the beginning of the creation of Symfony 2.0 (5-6 years now) to precisely allow the Bundles to self-configure, and you replied that it was very complicated as it was and that this was not the primary purpose of the Framework. So now that it's achievable and that it's from actuality, I'm more than enthusiastic :). Thanks again for your work!

I understand your argument on the evolution of the recipe, However in the case of a private package, the recipe will evolve in the package, But yes, in the case of removing the package, there could be a difference between the installed version and the remote version.

Before creating this issue, I was thinking of another possibility: allow to create a recipe in its own private repository, which would solve the problem, but there are other problems, and the approach of a private server is in this case more coherent.

However, why not do as the Composer VCS Repository to retrieve the recipe from the private package? That is, use the different Composer Drivers to retrieve the `manifest.json` file in the remote repository of the private package?

What do you think?

**Comment by francoispluchino** [comment]
You will find in this [comment](https://github.com/symfony/flex/issues/70#issuecomment-299915048) of the issue #70, the reasons and arguments to make Symfony Flex more flexible, in report with this proposal.

**Comment by fabpot** [comment]
Closing as private repos are now supported.

**Comment by francoispluchino** [comment]
@fabpot Thanks you for your affirmation, but how do private repositories work with Flex? I didn't find documentation, commits in Flex, or a new repository (for a private server).

Can we have more details for this new feature?

**Comment by fabpot** [comment]
It's on https://symfony.sh/ in the right column. Go here to register: https://github.com/apps/symfony-flex-server/installations/new

**Comment by natelenart** [comment]
@fabpot Is there any support for non-Github repositories?  My company is happy with our current setup - loading company-developed libraries/bundles from an internal Satis server pointing at an internal Bitbucket server.  From what I understand, with this setup, we're unable to create and use our own Flex recipes, is this correct?  We currently aren't using GitHub at all.

**Comment by hectorprats** [comment]
Just install a flex server, and write extra.symfony.endpoint in your composer.json with your private URL flex server

**Comment by c33s** [comment]
after microsoft aquired github its even more important to not require github as dependency.

**Comment by codedmonkey** [comment]
Can we get an update on this? I'm reading all kinds of stuff about it being discontinued but there's no word from the Core team about it?  Does https://github.com/apps/symfony-flex-server even do anything?

Whatever text @fabpot is referring to has been removed. I would like to say it's surprising that Symfony has a project that's so badly documented but honestly It's a recurring theme. Symfony needs to stop this crusade against small developers who can't do everything by themselves. That's not why I started contributing.

**Comment by fabpot** [comment]
Flex does not support private recipes anymore. Very few people were using them and in retrospect, we don't see that many valid use cases.

I won't comment about "badly documented" and "crusade against small developers" as I really don't understand what you're talking about. Anyway, have a nice day.

**Comment by andrew-demb** [comment]
@codedmonkey check this repository https://github.com/moay/server-for-symfony-flex for private server
